src/drivers/magnetometer/bosch/bmm350/BMM350.cpp

```
@@ -0,0 +1,757 @@
+}
+
+uint8_t BMM350::GetODR(int value)
+{
+	switch (value) {
+	case 0: return ODR_400HZ;
+
+	case 1: return ODR_200HZ;
+
+	case 2: return ODR_100HZ;
+
+	case 3: return ODR_50HZ;
+
+	case 4: return ODR_25HZ;
+
+	case 5: return ODR_12_5HZ;
+
+	case 6: return ODR_6_25HZ;
+
+	case 7: return ODR_3_125HZ;
+
+	case 8: return ODR_1_5625HZ;
+
+	default: return ODR_200HZ;
+	}
+}
+
+hrt_abstime BMM350::OdrToUs(uint8_t odr)
+{
+	switch (odr) {
+	case ODR_400HZ:
+		return 2500_us;
+
+	case ODR_200HZ:
+		return 5000_us;
+
+	case ODR_100HZ:
+		return 10000_us;
+
+	case ODR_50HZ:
+		return 20000_us;
+
+	case ODR_25HZ:
+		return 40000_us;
+
+	case ODR_12_5HZ:
+		return 80000_us;
+
+	case ODR_6_25HZ:
+		return 160000_us;
+
+	case ODR_3_125HZ:
+		return 320000_us;
+
+	case ODR_1_5625HZ:
+		return 640000_us;
+
+	default:
+		return 5000_us;
+	}
+}
+
+uint8_t BMM350::GetAVG(int value)
+{
+	switch (value) {
+	case 0: return AVG_NO_AVG;
+
+	case 1: return AVG_2;
+
+	case 2: return AVG_4;
+
+	case 3: return AVG_8;
+
+	default: return AVG_2;
+	}
+}
+
+
+int BMM350::probe()
+{
+	for (int i = 0; i < 3; i++) {
+		uint8_t chip_id;
+
+		if (PX4_OK == RegisterRead(Register::CHIP_ID, &chip_id)) {
+			PX4_DEBUG("CHIP_ID: 0x%02hhX", chip_id);
+
+			if (chip_id == chip_identification_number) {
+				return PX4_OK;
+			}
+		}
+	}
+
+	return PX4_ERROR;
+}
+
+void BMM350::RunImpl()
+{
+	const hrt_abstime now = hrt_absolute_time();
+	ParametersUpdate();
+
+	switch (_state) {
+	case STATE::RESET: {
+			if (RegisterWrite(Register::CMD, SOFT_RESET) == PX4_OK) {
+				_reset_timestamp = now;
+				_state = STATE::WAIT_FOR_RESET;
+				perf_count(_reset_perf);
+			}
+
+			ScheduleDelayed(10_ms);
+		}
+		break;
+
+	case STATE::WAIT_FOR_RESET: {
+			int ret = probe();
+
+			if (ret == PX4_OK) {
+				uint8_t status_0;
+				ret = RegisterRead(Register::PMU_STATUS_0, &status_0);
+
+				if (ret == PX4_OK && (status_0 & PWR_NORMAL) != 0) {
+					ret = PX4_ERROR;
```

We go into PX4_ERROR here but then on line 227 we just ignore the error.